File: 000201-000400/000279.完全平方数/279.py
import math

# 难度: 中等
class Solution:
    def numSquares(self, n: int) -> int:
        dp = [x for x in range(0, n+1)]
        for i in range(1, n+1):
            for j in range(0, int(math.sqrt(i))+1):
                dp[i] = min(dp[i], dp[i - j ** 2] + 1)
        return dp[n]


s = Solution()
print(s.numSquares(12))

File: 000201-000400/000279.完全平方数/279.js
/**
 * @param {number} n
 * @return {number}
 */
// 难度: 中等
var numSquares = function(n) {
    let dp = new Array(n + 1).fill(n+1)
    dp[0] = 0
    // console.log('dp = ', dp)
    for (let i = 1; i <= n; i ++) {
        for(let j = 0; j * j <= i; j ++) {
            dp[i] = Math.min(dp[i], dp[i - j * j] + 1)
        }
    }
    console.log('dp = ', dp)
    return dp[n]
};

// console.log(numSquares(1))
// console.log(numSquares(2))
// console.log(numSquares(3))
// console.log(numSquares(4))
// console.log(numSquares(5))
console.log(numSquares(12))
// console.log(numSquares(26))
// console.log(numSquares(18))
